### ISSUE TITLE
Runtime worsens over time with Adam

### ISSUE THREAD

**Issue by ndryden** [LLNL/lbann#56]
When using the Adam optimizer (`-q 3` in e.g. the DNN MNIST script), the time to complete each minibatch/epoch *increases* after some number of steps are completed.

An minimal way to easily reproduce this is to use a single process on Catalyst (you'll need to edit `run_lbann_dnn_mnist.sh` to do this), and train a network with one 2048-neuron fully-connected layer with no regularization. To get the timing, make sure you add the timer callback. After a few epochs, the average minibatch time will increase.

Some things I have tried:
* This does not seem to occur with Adagrad.
* This does seem dependent on the size of the layers (which makes sense-- layer layers mean more work).
* This is not an I/O artifact-- staging the data on the SSD does not eliminate the effect.
* This is not an OpenMP artifact-- commenting out the `#pragma omp parallel for` lines does not eliminate the effect.
* This is not due to the data not being aligned-- we always take the contiguous data path.
* This doesn't seem to be a compiler artifact-- I see it with both GCC 4.9.1 and 6.1.0.
* If you look at the timing breakdown in Tensorboard, there are noticeable increases in the update time for the fully-connected layer. Manually adding timing also indicates the update loop has increasing time.
* The size of the data being iterated over doesn't change.

The loop in Adam ([here](https://github.com/LLNL/lbann/blob/develop/src/optimizers/lbann_optimizer_adam.cpp#L127)) is pretty simple, so I'm not sure what's causing this. It may be there's something strange going on outside the optimizer that's impacting Adam's performance.

**Comment by ndryden** [comment]
(Brief note up front: I've made a change that lifts the learning rate multiply out of the loop, it's unrelated to this.)

I believe I have identified the issue. The line
```
x -= correction * m1 / (Sqrt(m2) + m_eps);
```
gets progressively slower. The `m2` variable (the running estimate of the second moment) gets quite small. Specifically, it sometimes becomes denormalized, which can drastically slow down floating point performance in hardware (100-1000x slower). We also sometimes see the GEMMs and GEMV in the backprop linearity slowing down, which would also be explained by an increasing proportion of denormalized values.

Introducing batch normalization seems to keep activations (and hence gradients) in a nicer range _some_ of the time, which avoids this problem. It helps for the default sigmoid activations a lot, but using ReLUs, we get significant numbers of denormalized values even in the first epoch.

I am going to look to try and identify the source of these denormalized values.
